### PR TITLE
Generic fetcher: Add lockfile

### PR DESCRIPTION
**Pull request by kosciCZ**
This commit adds custom lockfile for the generic feature, used in order to specify list of files to download and checksums to verify against.

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

### CODE REVIEW

File: cachi2/core/package_managers/generic.py [hunk 1]
```diff
@@ -1,13 +1,54 @@
...
+import yaml
+from pydantic import BaseModel, ValidationError, field_validator
+
+from cachi2.core.errors import PackageManagerError, PackageRejected
 from cachi2.core.models.input import Request
 from cachi2.core.models.output import RequestOutput
 from cachi2.core.models.sbom import Component
 from cachi2.core.rooted_path import RootedPath
 
+LOG = logging.getLogger(__name__)
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/generic.py:13]: nitpick: we usually define this as a lowercase variable.

File: cachi2/core/package_managers/generic.py [hunk 2]
```diff
@@ -1,13 +1,54 @@
...
 DEFAULT_LOCKFILE_NAME = "cachi2_generic.yaml"
 DEFAULT_DEPS_DIR = "deps/generic"
 
 
+class LockfileMetadata(BaseModel):
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/generic.py:18]: I wonder if it is not worth having the model definitions in a standalone file (even if they're small now). This way, we'd make this into a package, with at least two modules (one for the models, other for the resolving itself), and probably a better foundation for expanding later on.

File: cachi2/core/package_managers/generic.py [hunk 3]
```diff
@@ -1,13 +1,54 @@
...
+    """Defines format of the metadata section in the lockfile."""
+
+    version: Literal["1.0"]
+
+
+class LockfileArtifact(BaseModel):
+    """Defines format of a single artifact in the lockfile."""
+
+    download_url: str
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/generic.py:27]: Are we postponing adding the `purl` field?

File: tests/unit/package_managers/test_generic.py [hunk 4]
```diff
@@ -1,17 +1,53 @@
...
+
+LOCKFILE_NO_CHECKSUM_2 = """
+metadata:
+    version: '1.0'
+artifacts:
+    - download_url: https://redhat.com/artifact
+      checksums: {}
+"""
+
+LOCKFILE_NORMAL = """
```
Comment (brunoapimentel) [inline tests/unit/package_managers/test_generic.py:42]: nitpick: I'd call this `VALID_LOCKFILE`

File: cachi2/core/package_managers/generic.py (hunk 3 above)
Comment (kosciCZ) [reply cachi2/core/package_managers/generic.py:27]: Yes, as agreed upon [in the design](https://github.com/containerbuildsystem/cachi2/pull/652#discussion_r1777195032), the purl functionality will be available only for the maven artifacts.

File: cachi2/core/package_managers/generic.py (hunk 2 above)
Comment (kosciCZ) [reply cachi2/core/package_managers/generic.py:18]: Sounds good to me.

Comment (kosciCZ) [reply cachi2/core/package_managers/generic.py]: refactored into two modules

File: cachi2/core/package_managers/generic.py (hunk 1 above)
Comment (kosciCZ) [reply cachi2/core/package_managers/generic.py]: renamed

File: tests/unit/package_managers/test_generic.py [hunk 5]
```diff
@@ -39,10 +75,79 @@ def test_fetch_generic_source(
...
+
+@pytest.mark.parametrize(
+    ["lockfile", "expected_exception", "expected_err"],
+    [
+        pytest.param("{", PackageRejected, "yaml format is not correct", id="invalid_yaml"),
+        pytest.param(
+            LOCKFILE_WRONG_VERSION, PackageManagerError, "Input should be '1.0'", id="wrong_version"
+        ),
+        pytest.param(
+            LOCKFILE_NO_CHECKSUM_1, PackageManagerError, "Field required", id="no_checksum_1"
```
Comment (a-ovchinnikov) [inline tests/unit/package_managers/test_generic.py:97]: "no_checksum_1" does not really describe the test case. Would "checksum_field_is_missing" be a better desctiption?

**Review by a-ovchinnikov — APPROVED** [review]

File: cachi2/core/package_managers/generic/main.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+from cachi2.core.errors import PackageManagerError, PackageRejected
+from cachi2.core.models.input import Request
+from cachi2.core.models.output import RequestOutput
+from cachi2.core.models.sbom import Component
+from cachi2.core.package_managers.generic.models import GenericLockfileV1
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+DEFAULT_LOCKFILE_NAME = "cachi2_generic.yaml"
```
Comment (eskultety) [inline cachi2/core/package_managers/generic/main.py:14]: This will have to be renamed something else - it's part of the public API and we already know we're going to rebrand the project which means we'd have to rename the file as well -> BAD! Something like `generic_fetcher_lockfile.yaml` OR `generic_lockfile.yaml` might be good options, @brunoapimentel, @a-ovchinnikov  what do you think of the alternative names?

File: cachi2/core/package_managers/generic.py [hunk 7]
```diff
@@ -1,33 +0,0 @@
-from cachi2.core.errors import PackageRejected
-from cachi2.core.models.input import Request
-from cachi2.core.models.output import RequestOutput
-from cachi2.core.models.sbom import Component
-from cachi2.core.rooted_path import RootedPath
-
-DEFAULT_LOCKFILE_NAME = "cachi2_generic.yaml"
-DEFAULT_DEPS_DIR = "deps/generic"
```
Comment (eskultety) [inline cachi2/core/package_managers/generic.py:8]: If you're going to move code, then please do it in a separate commit.

File: cachi2/core/package_managers/generic/main.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        try:
+            lockfile_data = yaml.safe_load(f)
+        except yaml.YAMLError as e:
+            raise PackageRejected(
+                f"Cachi2 lockfile '{lockfile_path}' yaml format is not correct: {e}",
+                solution="Check correct 'yaml' syntax in the lockfile.",
+            )
+
+        try:
+            lockfile = GenericLockfileV1.model_validate(lockfile_data, context={})
```
Comment (eskultety) [inline cachi2/core/package_managers/generic/main.py:71]: nitpick: `GenericLockfileV1(**lockfile_data)` should take care of the validation as well.

File: cachi2/core/package_managers/generic.py (hunk 7 above)
Comment (kosciCZ) [reply cachi2/core/package_managers/generic.py:8]: Noted, will do so next time that's the case. Sorry about that.

File: cachi2/core/package_managers/generic/models.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+class LockfileArtifact(BaseModel):
+    """Defines format of a single artifact in the lockfile."""
+
+    download_url: str
+    target: Optional[str] = None
+    checksums: dict[str, str]
+
+    @field_validator("checksums")
+    @classmethod
+    def validate_checksums(cls, value: dict[str, str]) -> dict[str, str]:
```
Comment (eskultety) [inline cachi2/core/package_managers/generic/models.py:21]: nitpick: Should probably be called 'no_empty_checksums' as there isn't any validation happening here and based on the nature of the contents there can't even be any genuine validation since those strings have no semantic value.

File: cachi2/core/package_managers/generic/models.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+class LockfileMetadata(BaseModel):
+    """Defines format of the metadata section in the lockfile."""
+
+    version: Literal["1.0"]
+
+
+class LockfileArtifact(BaseModel):
+    """Defines format of a single artifact in the lockfile."""
+
+    download_url: str
```
Comment (eskultety) [inline cachi2/core/package_managers/generic/models.py:15]: Base model definitions should semantically be introduced in a separate commit to separate data types/interfaces from actual logic.

File: tests/unit/package_managers/test_generic.py [hunk 11]
```diff
@@ -39,10 +75,79 @@ def test_fetch_generic_source(
...
+
+@pytest.mark.parametrize(
+    ["lockfile", "expected_exception", "expected_err"],
+    [
+        pytest.param("{", PackageRejected, "yaml format is not correct", id="invalid_yaml"),
+        pytest.param(
+            LOCKFILE_WRONG_VERSION, PackageManagerError, "Input should be '1.0'", id="wrong_version"
+        ),
+        pytest.param(
+            LOCKFILE_CHECKSUM_MISSING, PackageManagerError, "Field required", id="no_checksum_1"
```
Comment (eskultety) [inline tests/unit/package_managers/test_generic.py:97]: 'no_checksum_1/2' isn't that much helpful, `no_checksum` and `checksum_empty` feel much more descriptive test IDs.

File: tests/unit/package_managers/test_generic.py [hunk 12]
```diff
@@ -39,10 +75,79 @@ def test_fetch_generic_source(
...
+    ["lockfile", "expected_lockfile"],
+    [
+        pytest.param(
+            LOCKFILE_VALID,
+            GenericLockfileV1.model_validate(
+                {
+                    "metadata": {"version": "1.0"},
+                    "artifacts": [
+                        {
+                            "download_url": "https://redhat.com/artifact",
```
Comment (eskultety) [inline tests/unit/package_managers/test_generic.py:133]: Don't need parametrize for a single test case.

File: tests/unit/package_managers/test_generic.py [hunk 13]
```diff
@@ -1,26 +1,62 @@
...
+    - download_url: https://redhat.com/artifact
+      checksums:
+        md5: 3a18656e1cea70504b905836dee14db0
+"""
+
+LOCKFILE_CHECKSUM_MISSING = """
+metadata:
+    version: '1.0'
+artifacts:
+    - download_url: https://redhat.com/artifact
```
Comment (eskultety) [inline tests/unit/package_managers/test_generic.py:31]: nitpick: Since the URLs are pretty much just placeholders in unit tests, let's stay "unaffiliated" and try to come up with some 'foobarized' vendor-agnostic version of it :) (there should be other examples in the code base already)

**Review by eskultety — COMMENTED** [review]
Looking good. I pretty much only raised nitpicks, but since I requested the changes to be split logically into more commits I raised these anyway since you'd be touching the code in the process of doing that.

File: cachi2/core/package_managers/generic/main.py (hunk 6 above)
Comment (brunoapimentel) [reply cachi2/core/package_managers/generic/main.py:14]: Either what you suggested, or just `generic-artifacts.yaml`. In any case, we don't need to settle on a name right now, since this is still under the experimental flag.

Comment (eskultety) [reply cachi2/core/package_managers/generic/main.py:14]: The development could move fast and so I'd like to at least get rid of the project reference ASAP otherwise we forget.

Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: Renamed to `generic_lockfile.yaml`

File: cachi2/core/package_managers/generic/main.py (hunk 8 above)
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: True, but I went with what was already in the codebase, because the [RPM package manager does basically the same thing](https://github.com/containerbuildsystem/cachi2/blob/d47d0516dbc4d3ef6e54d5537be720afc8eb1d56/cachi2/core/package_managers/rpm/main.py#L275).

File: tests/unit/package_managers/test_generic.py (hunk 12 above)
Comment (kosciCZ) [reply tests/unit/package_managers/test_generic.py]: I know I don't need to, but this feels cleaner, since it separates the data from the test logic. I've received the opposite instructions on the previous [PR](https://github.com/containerbuildsystem/cachi2/pull/671#discussion_r1787518942). If this is a blocker I don't mind changing it.

**Review by brunoapimentel — APPROVED** [review]

File: cachi2/core/package_managers/generic/main.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+            solution=(
+                f"Make sure your repository has cachi2 generic lockfile '{DEFAULT_LOCKFILE_NAME}' checked in "
+                "to the repository."
+            ),
+        )
+
+    log.info(f"Reading generic lockfile: {lockfile_path}")
+    lockfile = _load_lockfile(lockfile_path)
+    for artifact in lockfile.artifacts:
+        log.info(f"Resolving artifact: {artifact.download_url}")
```
Comment (eskultety) [inline cachi2/core/package_managers/generic/main.py:51]: This one should be a debug-level log rather than info.

**Review by eskultety — APPROVED** [review]
One tiny comment on reducing the logging noise, adding my ACK regardless.

File: cachi2/core/package_managers/generic/main.py (hunk 14 above)
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: Fixed.

**Comment by kosciCZ** [comment]
Thank you for your review and approvals! Can someone please merge?
